### ISSUE TITLE
Replacing runtime framework assembly to add own API implementations

### ISSUE THREAD

**Issue by a-ctor** [dotnet/runtime#37695]
In my company we have few external libraries that depend on .NET Framework specific API, which we now want to use under .NET Core. Sadly, we cannot re-target the assembly onto .NET Core (license problems), which is why we have to reference the .NET Framework assemblies directly from .NET Core.

In .NET Core there are assemblies like `System.Web` that forward classes to the .NET Core implementations (`HttpUtility`, for example, is forwarded from `System.Web.dll` to `System.Web.HttpUtility.dll`). This includes only the .NET Core API surface.

The problem is that the external libraries bind to `System.Web.dll`, which is missing some of the APIs the library needs. We now want to provide our own implementations for the APIs that are required by the external libraries. Since we cannot re-target the library we want to replace the framework `System.Web.dll` with our own version that includes additional type forwards to our own implementation of those APIs.

We have tried creating our own version of `System.Web.dll` (public signed) and loading that at the start of the program, but this doesn't work. It will read its assembly name and load the version located in the framework folder.

Is there a way to replace such a framework assembly without creating a custom SDK/framework?

**Comment by jkotas** [comment]
This should help:
- Make sure that your custom build has higher version number than the inbox version of the library
- Reference the library in your app .csproj to make it part of the app construction. The library needs to be recorded in `*.deps.json` file to override the inbox version.

**Comment by a-ctor** [comment]
Thank you @jkotas for you help. That worked perfectly :)

**Comment by a-ctor** [comment]
Sadly, I have made an error when testing and tampered with my SDK for @jkotas solution to work. With a clean install the version push with the direct reference is not enough to provide a custom assembly that overrides a framework assembly.

In my tests I tried to create a custom version of `System.Web.dll` and use it instead of the framework version. The assembly version is `4.0.1.0`, which will be preferred over the framework version `4.0.0.0`. With a clean SDK the build will fail as two versions of  `System.Web.dll` are found (see image below). Although the warning states that the compiler will use version `4.0.1.0` it passes both libraries as references to the compiler, which causes CS1703.

![image](https://user-images.githubusercontent.com/5082673/85837429-43e1b980-b798-11ea-80e5-5d5b83d0571c.png)

It is possible to get rid of this compiler error by removing (renaming works too) the reference assembly for `System.Web.dll` (`C:\Program Files\dotnet\packs\Microsoft.NETCore.App.Ref\5.0.0-preview.6.20305.6\ref\net5.0\System.Web.dll`). This allows the project to compile but at runtime the framework assembly is still preferred over the custom one. The assemblies `.deps.json` contains an entry for `System.Web./4.0.1.0` but the framework assembly still has higher precedence.

The only way to fix this problem is to remove the `System.Web.dll` from the `deps.json` in the shared framework folder (`C:\Program Files\dotnet\shared\Microsoft.NETCore.App\5.0.0-preview.6.20305.6\Microsoft.NETCore.App.deps.json`). This, of course, requires providing a custom runtime, which is what we wanted to prevent.

**Comment by a-ctor** [comment]
I have created a test repo with my current test setup: https://github.com/a-ctor/ReplaceFrameworkDllTest

Any help would be much appreciated :)

**Comment by jkotas** [comment]
> It is possible to get rid of this compiler error by removing (renaming works too) the reference assembly

You can remote it programmatically by adding something like this to test.csproj:

```
  <Target Name="RemoveSystemWeb" AfterTargets="ResolveTargetingPackAssets" BeforeTargets="ResolvePackageFileConflicts">
    <ItemGroup>
      <ReferencesToExclude Include="@(Reference)" Condition="'%(FileName)' == 'System.Web'" />
      <Reference Remove="@(ReferencesToExclude)" />
    </ItemGroup>
  </Target>
```

> The assemblies .deps.json contains an entry for System.Web./4.0.1.0 but the framework assembly still has higher precedence.

The entry in test.deps.json needs to have file and/or assembly versions like this for your local assembly to have higher precedence:

```
        "runtime": {
          "System.Web.dll": {
            "assemblyVersion": "9.9.9.9",
            "fileVersion": "9.9.9.9"
          }
        }
```

I am not sure what is the best way to convince the build tooling to emit these versions for your local app assemblies. You try  debugging https://github.com/dotnet/sdk/search?q=GenerateDepsFile&unscoped_q=GenerateDepsFile to figure it out.

**Comment by a-ctor** [comment]
Thank you very much @jkotas for your help and quick response.

I didn't find a way to convince the build tooling to emit versions when using `ProjectReference` items. But I noticed that the versions were generated for Nuget packages and `Reference` items. 


So I created a Nuget package with my `System.Web` assembly and the your build script. Reference that from a project and everything works like a charm.
